### ISSUE TITLE
Move MPI-1 deprecated functions to new "Removed Interfaces" chapter.

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#303]
**Originally by _ftillier_ on 2011-12-02 13:17:13 -0600**

---

This ticket implements part B of ticket #278:

Remove the following deprecated functions from the Deprecated chapter: 
- MPI_TYPE_HVECTOR 
- MPI_TYPE_HINDEXED 
- MPI_TYPE_STRUCT 
- MPI_ADDRESS 
- MPI_TYPE_EXTENT 
- MPI_TYPE_LB 
- MPI_TYPE_UB 
- MPI_ERRHANDLER_CREATE (and its function pointer typedef) 
- MPI_ERRHANDLER_SET 
- MPI_ERRHANDLER_GET 

Brian brought up an annoying (but valid) point: if we actually remove these functions from the spec, then MPI implementations cannot include them because they will violate the MPI_\* namespace. Gaaahh!! That kinda defeats the rationale that MPI implementations can still carry these functions for a while if they want to.

So Brian made a good suggestion here: what if we just move these functions to a new, "These functions have been deleted" chapter? GNU Autoconf does this; it has a listing of deprecated / removed macros with a 1-sentence description of what it was replaced with (or a short explanation if nothing replaced it). This would allow the symbols/typedefs/constants/whatever to still be listed in the MPI spec, but they're now optional for implementations. And there's even a helpful 1-line pointer saying "MPI_ADDRESS was replaced with MPI_GET_ADDRESS."

I think that this is a fine solution. 

Rolf (and Pavan) wants to amend this to just move the exact text from the Deprecated chapter to the Deleted chapter (i.e., keep the prototypes and any minimal descriptions that are there now). I'd prefer to not keep the prototypes; make it a little hard to keep using the old functions.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-12-02 13:19:14 -0600**

---

WikiFormatting > me.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-12-12 18:42:19 -0600**

---

**Attachment added:** [`ticket-303.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/303/ticket-303.pdf) (2297.8 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2011-12-12 19:48:34 -0600**

---

Frankly, I don't understand the haste in deleting these functions.  Isn't deprecated good enough?  Removing them from the examples makes sense, but removing deprecated functions breaks existing applications, something that I don't think MPI should do.  Applications in scientific computing often live _much_ longer than in some other areas; many applications are multiple decades old.  Non-backward-compatible changes to the relevant standards should have similar lifetimes.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-12-12 20:38:26 -0600**

---

Replying to gropp:

> Frankly, I don't understand the haste in deleting these functions.  Isn't deprecated good enough?  Removing them from the examples makes sense, but removing deprecated functions breaks existing applications, something that I don't think MPI should do.  Applications in scientific computing often live _much_ longer than in some other areas; many applications are multiple decades old.  Non-backward-compatible changes to the relevant standards should have similar lifetimes.

One of the reasons for moving these functions to a "Removed" chapter, rather than deleting them outright, was to preserve backward compatibility.  There is nothing preventing a library from providing MPI-2.2 support as well as MPI-3.0 support.  MPI-3.0 implementations are thus given the choice whether to support these deprecated functions, a choice that would presumably be driven by customer demand.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2011-12-13 14:08:56 -0600**

---

FWIW, I seem to recall that there was a straw vote about whether to do this or not at the meeting, and the Forum was strongly in favor of moving deprecated stuff to a "deleted" chapter so that an implementation could still choose to support them if they want to.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-12-20 15:22:10 -0600**

---

Retracting comment, as the straw vote results were for C++, not the MPI-1 functionality...

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-10 23:03:40 -0600**

---

Notes from the Jan 2012 Forum meeting:
- p497 line 18: "The following MPI bindings were deprecated as of MPI-2.0 and are removed in MPI-3.0."
- p497 and p498: Also change "Removed MPI-1" to "Removed MPI" in all cases
- Will also need to remove the C++ constants...?

This ticket depends on #278, so it cannot be voted on until #278 is voted on.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-01-17 10:43:54 -0600**

---

I propose to withdraw this ticket for Fortran reasons:
- Fortran applications need changes in the data declarations (64bit INTEGER) and mpif.h and application-subroutines may still not use argument checking --> high risks
- Fortran rule for “deprecated” means that features should not be used in new applications, but they are still available for existing applications the next ~20 years.[[BR]]
  Example: Redundant features in F77, removed in F95:  ASSIGN, H, PAUSE, ...
  See also http://h71000.www7.hp.com/doc/82final/6324/6324pro_059.html A.1

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-18 10:21:41 -0600**

---

I respectfully disagree.  :-)
- We're not Fortran.  
- These interfaces have been deprecated since 1996.  So yes, it's _only_ been 15 years -- but that's enough.
- We're not forcing implementations to remove the functions; we're just saying that they don't have to have them any more.
- The way I see it, applications will have three options:
  1. Upgrade their code.
  2. Do not upgrade their code and stay with older versions of their MPI implementations.
  3. Do not upgrade their code and petition their MPI implementors to keep the MPI-1 deprecated functions.

I believe that these are reasonable options.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-05-25 08:32:58 -0500**

---

There are strong reasons to reject this ticket #303: 
- There is no way to directly substitute the old MPI_LB + MPI_UB + MPI_STRUCT by new MPI_TYPE_RESIZE  or  MPI_TYPE_STRUCT + MPI_TYPE_RESIZE or MPI_TYPE_CREATE_SUBARRAY.[[BR]]
  
  Therefore there is a **high risk** to break user applications and **high costs** to verify the changes. 
- There is no way to directly substitute the Fortran deprecated routines by the new MPI-2.0 routines because the new routines have long integer instead of short integer. If such variables are part of data-structures then new 8-bit-alignment problems can arise.[[BR]]
  
  Therefore again is a **high risk** to break user applications and **high costs** to verify the changes.
- Only the C routines (but not MPI_LB und MPI_UB) can be directly substituted.[[BR]]
  
  The forum is responsible to look at all aspects, and not only at the simple aspects.
  MPI_LB and MPI_UB to build multi-dimensional subarrays and all the Fortran routines should not be ignored. 
- Applications could not be updated until most MPI libraries were MPI-2.2 compliant. This was many years after the official standardization in 1997.[[BR]]
  
  From a user's perspective, the new MPI-2.0 substitutes are available in all MPI libraries for only a few years now.

Therefore "deprecated" should mainly mean that users should not continue to newly use these routines. And therefore they should be kept as part of the MPI-3.0 standard.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-17 09:43:56 -0500**

---

Committed changes to binding chapter.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-17 21:33:33 -0500**

---

Committed changes to
- chap-terms
- chap-deprecated
- chap-removed
- chap-appLang
  in svn r1382.

Missing chapters are chap-datatype and chap-inquiry.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2012-07-18 09:26:04 -0500**

---

working to apply 303 to datatypes...

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 09:53:34 -0500**

---

Appears to be fully committed. Moving to "Waiting for PDF reviews".

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2012-07-18 11:31:20 -0500**

---

applied to datatypes chapter as of r1398

**Comment by mpiforumbot** [comment]
**Originally by _schulzm_ on 2012-07-18 14:31:15 -0500**

---

Applied to Chapter 8, MPI Environmental Management

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2012-07-19 12:34:25 -0500**

---

The context chapter has a reference to a section that was removed as part of this ticket (the last sentence of the the second paragraph before 6.7.1).  Has the "discussion of the address size problem" been moved somewhere else that can be referenced, or should the sentence referencing the deleted section be removed?

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2012-07-19 13:40:42 -0500**

---

This text needs to be rewritten, since the entire paragraph was built around the "address size problem".  I'll take care of it.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-21 04:48:05 -0500**

---

The description did not tell that this ticket also removes three constants that were introduced in MPI-2.0 (and therefore not part of the deprecated MPI-1.1 routines):
- MPI_COMBINER_HVECTOR_INTEGER
- MPI_COMBINER_HINDEXED_INTEGER
- MPI_COMBINER_STRUCT_INTEGER

It is meaningful to also remove these constants, but it should be correctly documented.
Therefore, I'll list these in the list of deprecated and removed constructs in the Terms chapter with "since deprecared" "-" and "since removed" "MPI-3.0".
